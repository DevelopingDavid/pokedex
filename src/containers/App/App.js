import React, { Component } from 'react';
import { connect } from 'react-redux';
import { grabAllPokemon } from '../../actions';
import CardContainer from '../CardContainer/CardContianer';
import { Switch, Route } from 'react-router-dom';
import Popup from '../../components/Popup/Popup';

export class App extends Component {
  constructor() {
    super()
    this.state = {
      loading: true
    }
  }

  componentDidMount() {
    this.fetchGenOne();
  }

  fetchGenOne = async () => {
    const response = await fetch(`https://pokeapi.co/api/v2/generation/1/`);
    const data = await response.json();
    const resolved = await this.fetchInfo(data.pokemon_species);
    console.log('in')
    this.cleanData(resolved);
  }
  
  fetchInfo = async (pokemon) => {
    let pokemonInfo = pokemon.map(async pokemon => {
      const id = pokemon.url.substring(42);
      const response = await fetch(`https://pokeapi.co/api/v2/pokemon/${id}`);
      const data = await response.json();
      return data;
    });
    return Promise.all(pokemonInfo)
  }
  
  cleanData = (pokemon) => {
    let cleaned = pokemon.map(pokemon => ({
      name: pokemon.name,
      sprite: pokemon.sprites.front_default,
      height: pokemon.height,
      weight: pokemon.weight,
      types: pokemon.types,
      dexNumber: pokemon.id
    }));
    const sorted = cleaned.sort((a, b) => {
     return a.dexNumber - b.dexNumber
    });
    this.props.grabPokemon(sorted);
    this.setState({ loading: false })
  }

  render() {
    return (
      <section className="App">
        {
          !this.state.loading &&
          <div>
          <header>
          <h1>Gen 1 Pok√©dex</h1>
          </header>
          <Switch>
            <Route path="/" exact component={CardContainer} />
            <Route path="/pokemon/:id" render={({ match }) => {
              const { id } = match.params;
              const currentPokemon = this.props.pokemon.find(pokemon => pokemon.dexNumber == id);
              return <Popup id={currentPokemon.dexNumber} currentPokemon={currentPokemon} />
            }} />
          </Switch>
          </div>
        }

      </section>
    );
  }
}

export const mapStateToProps = (state) => ({
  pokemon: state.pokemon
});

export const mapDispatchToProps = (dispatch) => ({
  grabPokemon: (pokemon) => dispatch(grabAllPokemon(pokemon))
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
