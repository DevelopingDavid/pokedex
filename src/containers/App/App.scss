.App {
  background: #313131;
  min-height: 850px;
}

.logo {
  width: 20%;
  height: auto;
  margin: 10px;
}