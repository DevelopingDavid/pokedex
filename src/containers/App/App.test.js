import React from 'react';
import { shallow } from 'enzyme';
import { App } from './App';
let mockPokemon = [{
  height: 22, 
  id: 149, 
  name: "dragonite",
  sprites: { front_default: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/149.png" },
  weight: 2100
}];

describe('App', () => {
  let wrapper;
  let instance;
  beforeEach(() => {
    wrapper = shallow(
      <App grabPokemon={jest.fn()}/>
    )
    instance = wrapper.instance();
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have default state', () => {
    expect(wrapper.state()).toEqual({
      loading: true
    });
  });

  it('should call upon fetchGenOne on componentDidMount', () => {
    const spy = jest.spyOn(instance, 'fetchGenOne');

    instance.componentDidMount();

    expect(spy).toHaveBeenCalled();
  });

  it('should fetch pokemon upon invoking fetchGenOne', async () => {
    const mockSuccessResponse = {};
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise,
    });

    jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise);

    instance.fetchGenOne();

    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith('https://pokeapi.co/api/v2/generation/1/');
  });
  
  it('should call upon sortData invoking cleanData', () => {
    let spy = jest.spyOn(instance, 'sortData');

    instance.cleanData(mockPokemon);

    expect(spy).toHaveBeenCalled();
  });

  it('should setState upon invoking cleanData', () => {
    instance.cleanData(mockPokemon);
    expect(wrapper.state('loading')).toEqual(false)
  });
});