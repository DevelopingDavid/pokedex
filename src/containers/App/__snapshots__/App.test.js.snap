// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<section
  className="App"
>
  <header>
    <img
      alt="logo"
      className="logo"
      src="logo.png"
    />
  </header>
</section>
`;
